File: 74.search-a-2-d-matrix.py
#
# @lc app=leetcode id=74 lang=python3
#
# [74] Search a 2D Matrix
#


# @lc code=start
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # run two binary search
        # first search which row
        # then search col
        firstNums = [row[0] for row in matrix]
        # get which row
        l, r = 0, len(firstNums) - 1
        while l <= r:
            mid = (l + r) // 2
            midNum = firstNums[mid]
            if midNum == target:
                return True
            if midNum > target:
                r = mid - 1
            else:
                l = mid + 1
        rowIdx = r
        if rowIdx < 0:
            return False
        # search in rows
        row = matrix[rowIdx]
        l, r = 0, len(row) - 1
        while l <= r:
            mid = (l + r) // 2
            if row[mid] == target:
                return True
            if row[mid] > target:
                r = mid - 1
            else:
                l = mid + 1
        return False


# @lc code=end

File: 74.search-a-2-d-matrix.go
/*
 * @lc app=leetcode id=74 lang=golang
 *
 * [74] Search a 2D Matrix
 */

// @lc code=start
package main

/*
1. find which row -> binary search to find the right most item that is lower or equal than target
2. find if the number is presented in that row
*/
func searchMatrix(matrix [][]int, target int) bool {
	n := len(matrix)
	l, r := 0, n-1
	for l <= r {
		m := (l + r) / 2
		if matrix[m][0] == target {
			return true
		}
		if matrix[m][0] < target {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	if r == -1 {
		return false
	}
	row := matrix[r]
	m := len(row)
	l, r = 0, m-1
	for l <= r {
		m := (l + r) / 2
		if row[m] == target {
			return true
		}
		if row[m] < target {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	return false
}

// @lc code=end
